go-support/cloudstate/eventsourced.go

```diff
@@ -0,0 +1,400 @@
+		esh.methodCache[cacheKey] = method
+	}
+
+	// build the input arguments for the method we're about to call
+	inputs := make([]reflect.Value, method.Type.NumIn())
+	inputs[0] = entityValue
+	inputs[1] = reflect.ValueOf(server.Context())
+
+	// create a zero-value for the type of the
+	// message we call the method with
+	arg1 := method.Type.In(2)
+	ptr := false
+	for arg1.Kind() == reflect.Ptr {
+		ptr = true
+		arg1 = arg1.Elem()
+	}
+	var msg proto.Message
+	if ptr {
+		msg = reflect.New(arg1).Interface().(proto.Message)
+	} else {
+		msg = reflect.Zero(arg1).Interface().(proto.Message)
+	}
+	if proto.Unmarshal(cmd.GetPayload().GetValue(), msg) != nil {
+		log.Fatalf("failed to unmarshal") // FIXME
+	}
+
+	inputs[2] = reflect.ValueOf(msg)
+	// call it
+	called := method.Func.Call(inputs)
+	// The gRPC implementation returns the rpc return method
+	// and an error as a second return value.
+	errReturned := called[1]
+	if errReturned.Interface() != nil && errReturned.Type().Name() == "error" { // FIXME: looks ugly
+		// TCK says: FIXME Expects entity.Failure, but gets lientAction.Action.Failure(Failure(commandId, msg)))
+		failure := &protocol.Failure{
+			CommandId:   cmd.GetId(),
+			Description: errReturned.Interface().(error).Error(),
+		}
+		failedToSend := sendClientActionFailure(failure, server)
+		if failedToSend != nil {
+			panic(failedToSend) // TODO: don't panic
+		}
+		return
+	}
+	// the reply
+	callReply, ok := called[0].Interface().(proto.Message)
+	if !ok {
+		log.Fatalf("called return value at index 0 is no proto.Message")
+	}
+	typeUrl := fmt.Sprintf("%s/%s", protoAnyBase, proto.MessageName(callReply))
+	marshal, err := proto.Marshal(callReply)
+	if err != nil {
+		log.Fatalf("unable to Marshal command reply")
+	}
+
+	// emitted events
+	events := esh.marshalEventsTo(entityValue)
+	esh.handleEvents(entityValue, events)
+	err = sendEventSourcedReply(&protocol.EventSourcedReply{
+		CommandId: cmd.GetId(),
+		ClientAction: &protocol.ClientAction{
+			Action: &protocol.ClientAction_Reply{
+				Reply: &protocol.Reply{
+					Payload: &any.Any{
+						TypeUrl: typeUrl,
+						Value:   marshal,
+					},
+				},
+			},
+		},
+		Events: events,
+	}, server)
+	if err != nil {
+		log.Fatalf("unable to send")
+	}
+}
+
+// marshalEventsTo receives the events emitted through the handling of a command
+// and marshals them to the event serialized form.
+func (esh *EventSourcedHandler) marshalEventsTo(entityValue reflect.Value) []*any.Any {
+	events := make([]*any.Any, 0)
+	if emitter, ok := entityValue.Interface().(EventEmitter); ok {
+		for _, evt := range emitter.Events() {
+			// TODO: protobufs are expected here, but CloudState supports other formats
+			message, ok := evt.(proto.Message)
+			if !ok {
+				log.Fatalf("got a non-proto message as event")
+			}
+			marshal, err := proto.Marshal(message)
+			if err != nil {
+				log.Fatalf("unable to Marshal")
+			}
+			events = append(events,
```

Comment:
>Did you have any concerns regarding the order in which events are emitted
@akramtexas Actually I'm not. This should be completely in sequence of what is stated instruction wise. Since theres is, at the first call shortly discussed, no concurrent data path involved, the EventSourcedHandler has at all times full control over the path how events are flowing around. Do you see any races here?

Reply:
@marcellanz Here, too, you have removed my concern. Looks good. For me, the final word on concurrency (in `Go`!) is the fine book named _“Concurrency in Go”_ by Katherine Cox-Buday, where she has a helpful tip toward the end of the book. FWIW, like so:
```
“In Go 1.1, a -race flag was added as a flag for most go commands:
  $ go test -race mypkg    # test the package
  $ go run -race mysrc.go  # compile and run the program
  $ go build -race mycmd   # build the command
  $ go install -race mypkg # install the package
If you’re a developer and all you need is a more reliable way to detect race conditions, this is really all you need to know. One caveat of using the race detector is that the algorithm will only find races that are contained in code that is exercised. For this reason, the Go team recommends running a build of your application built with the race flag under real-world load. This increases the probability of finding races by virtue of increasing the probability that more code is exercised.”
```

Reply:
@akramtexas yeah the `-race` flag is of great help for detecting races in code. As a precaution I have been using this flag with the [`go-support/build/run_go_test_in_docker.sh`](https://github.com/cloudstateio/cloudstate/pull/103/commits/c1afd6ac4ff0423c2f74c7c89db37d7a12f59356#diff-b92d99f75509105ea6d3ff4d474113bdR7) script

`go test -count 1 -race ./...`

We might even run the TCK with the flag enabled to get covered here.

Reply:
@marcellanz Good to hear about your routine use of the `-race` flag 👍 And yes, I support running the TCK with the flag enabled to get covered in this area of `Go` code, provided that it does not create any (infrastructural) burden.